packages/ckeditor5-table/src/tableutils.js

```diff
@@ -132,37 +137,52 @@ export default class TableUtils extends Plugin {
 			}
 
 			// Inserting at the end and at the beginning of a table doesn't require to calculate anything special.
-			if ( insertAt === 0 || insertAt === table.childCount ) {
-				createEmptyRows( writer, table, insertAt, rowsToInsert, this.getColumns( table ) );
+			if ( copyStructureFrom < 0 && ( insertAt === 0 || insertAt === rows ) ) {
+				createEmptyRows( writer, table, insertAt, rowsToInsert, columns );
 
 				return;
 			}
 
-			// Iterate over all rows above inserted rows in order to check for rowspanned cells.
-			const tableIterator = new TableWalker( table, { endRow: insertAt } );
+			// Iterate over all rows above inserted rows in order to check for row-spanned cells.
+			const tableIterator = new TableWalker( table, { endRow: Math.max( insertAt, copyStructureFrom ) } );
 
 			// Will hold number of cells needed to insert in created rows.
-			// The number might be different then table cell width when there are rowspanned cells.
-			let cellsToInsert = 0;
+			// The number might be different then table cell width when there are row-spanned cells.
+			let cellsToInsert = columns;
+
+			// Store spans of the reference row to reproduce it's structure. This array is column number indexed.
+			const columnsSpans = new Array( columns ).fill( 1 );
 
-			for ( const { row, rowspan, colspan, cell } of tableIterator ) {
-				const isBeforeInsertedRow = row < insertAt;
-				const overlapsInsertedRow = row + rowspan > insertAt;
+			for ( const { row, column, rowspan, colspan, cell } of tableIterator ) {
+				const lastCellRow = row + rowspan - 1;
 
-				if ( isBeforeInsertedRow && overlapsInsertedRow ) {
+				if ( row < insertAt && insertAt <= lastCellRow ) {
 					// This cell overlaps inserted rows so we need to expand it further.
 					writer.setAttribute( 'rowspan', rowspan + rowsToInsert, cell );
+
+					// Mark this column as spanned from the rows above so no new cell will be needed.
+					columnsSpans[ column ] = -colspan;
+					cellsToInsert -= colspan;
+				} else if ( row <= copyStructureFrom && copyStructureFrom <= lastCellRow ) {
+					columnsSpans[ column ] = colspan;
+					cellsToInsert -= colspan - 1;
 				}
+			}
 
-				// Calculate how many cells to insert based on the width of cells in a row at insert position.
-				// It might be lower then table width as some cells might overlaps inserted row.
-				// In the table above the cell 'a' overlaps inserted row so only two empty cells are need to be created.
-				if ( row === insertAt ) {
-					cellsToInsert += colspan;
```

Comment:
Previously it was additive calculations. Now everything is susbstractive. It looks like it require more calculations.

ATM I'm thinking about simpler switch which would allow to:

```javascript
cellsToInsert += useSingleColumns ? colspan : 1;
```

But I'm just "thinking" aloud here - I'm not sure if this would be the best way because we need to copy some attributes.